src/v/kafka/server/group.cc

```diff
@@ -1875,16 +1874,28 @@ group::prepare_tx(cluster::prepare_group_tx_request r) {
       raft::replicate_options(raft::consistency_level::quorum_ack));
 
     if (!e) {
+        // Prev logic delete tx from volatile before replication. Sp we should
+        // do deletion before all logic with step down
+        _volatile_txs.erase(r.pid);
+
         // Situation: replication has passed but the replicate method returns
         // error - _volatile_txs is empty, _prepared_txs is empty so when a
         // fetch happens the group thinks there is no active transaction and
         // doesn't block while it should. So we do step_down to give next leader
         // reply log and understand status of transaction
         co_await _partition->raft()->step_down();
+
         _expiration_info.erase(r.pid);
+
         co_return make_prepare_tx_reply(cluster::tx_errc::unknown_server_error);
     }
 
+    // We move deletion tx from volatile after replication, because
+    // fetch_offset can see internal state when redpanda has ongoing tx, but it
+    // is not inside internal maps. So we need do moving tx between maps in
+    // atomic section
+    _volatile_txs.erase(r.pid);
```

Comment:
probably good to do the same in abort_tx()? (similar logic).

Reply:
I think not
In abort we delete it before replicate, because wanna prevent prepare request. If we do it after replicate stage we broke this logic
Need to chat with @rystsov

Reply:
`abort_tx` should be fine, we may not block the fetch offsets call because an inflight abort doesn't interfere with it

Reply:
> `abort_tx` should be fine, we may not block the fetch offsets call because an inflight abort doesn't interfere with it

But it is another problem from comment in abort
```
    // preventing prepare and replicte once we
    // know we're going to abort tx and abandon pid
```